### PR TITLE
Update Join Cardinality Estimation for Non-Int and Non-Numeric columns

### PR DESCRIPTION
**Pull request by vraghavan78**
Histogram intersection depends on the value of the bucket boundaries. For datatypes like text, varchar, etc. Orca currently uses a hash function. This function is useful for equality with singleton buckets but nothing more. So the join stats computation based on histogram intersection is totally bogus. I now modified it into a NDV (number of distinct values) based estimation. In this PR I also did some test code refactoring.

### CODE REVIEW

File: server/src/unittest/dxl/statistics/CCardinalityTestUtils.cpp [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+	INT iLower,
+	INT iUpper,
+	CDouble dFrequency,
+	CDouble dDistinct
+	)
+{
+	CPoint *ppLower = CTestUtils::PpointInt4(pmp, iLower);
+	CPoint *ppUpper = CTestUtils::PpointInt4(pmp, iUpper);
+
+	BOOL fUpperClosed = false;
```
Comment (oarap) [inline server/src/unittest/dxl/statistics/CCardinalityTestUtils.cpp]: So we consider it is closed when its lower bound is closed? [x,y) unless x=y?

File: server/src/unittest/dxl/statistics/CCardinalityTestUtils.cpp [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+	if (ppLower->FEqual(ppUpper))
+	{
+		fUpperClosed = true;
+	}
+	CBucket *pbucket = GPOS_NEW(pmp) CBucket
+									(
```
Comment (oarap) [inline server/src/unittest/dxl/statistics/CCardinalityTestUtils.cpp]: looks ugly, remove some tabs

File: server/src/unittest/dxl/statistics/CCardinalityTestUtils.cpp [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+	BOOL fValue,
+	CDouble dFrequency
+	)
+{
+	CPoint *ppLower = CTestUtils::PpointBool(pmp, fValue);
+
+	// lower bound is also upper bound
+	ppLower->AddRef();
+	CBucket *pbucket = GPOS_NEW(pmp) CBucket
+									(
```
Comment (oarap) [inline server/src/unittest/dxl/statistics/CCardinalityTestUtils.cpp]: same here, delete some tabs

File: libnaucrates/src/statistics/CHistogram.cpp [hunk 4]
```diff
@@ -1318,6 +1319,51 @@ CHistogram::PhistJoinEquality
...
+		{
+			dNDV2 = dNDV2 - 1.0;
+			dFreqRemain2 = dFreqRemain2 - this->DNullFreq();
+		}
+
+		// the estimated final number of distinct value for the join is the minimum of the non-null
+		// distinct values of the two inputs.
+		// Frequency follows the principle of used to estimate join scaling factor -- based on
+		// the maximum NDV of the two inputs
+		dDistinctRemain = std::min(dNDV1, dNDV2);
```
Comment (hsyuan) [inline libnaucrates/src/statistics/CHistogram.cpp]: It is minimum, not maximum

Comment (hsyuan) [inline libnaucrates/src/statistics/CHistogram.cpp]: Is it OK to be -1?

File: libnaucrates/src/statistics/CHistogram.cpp [hunk 5]
```diff
@@ -1318,6 +1318,51 @@ CHistogram::PhistJoinEquality
...
+			dNDV1 = dNDV1 - 1.0;
+			dFreqRemain1 = dFreqRemain1 - this->DNullFreq();
+		}
+
+		CDouble dNDV2 = phist->DDistinct();
+		CDouble dFreqRemain2 = phist->DFrequency();
+		if (CStatistics::DEpsilon < phist->DNullFreq())
+		{
+			dNDV2 = dNDV2 - 1.0;
+			dFreqRemain2 = dFreqRemain2 - this->DNullFreq();
```
Comment (vraghavan78) [inline libnaucrates/src/statistics/CHistogram.cpp]: good catch. cut copy paste error

File: libnaucrates/src/statistics/CHistogram.cpp (hunk 4 above)
Comment (vraghavan78) [inline libnaucrates/src/statistics/CHistogram.cpp]: no. but good point

File: libnaucrates/src/statistics/CHistogramUtils.cpp [hunk 6]
```diff
@@ -0,0 +1,661 @@
...
+#include "gpopt/base/CColumnFactory.h"
+
+#include "gpopt/engine/CStatisticsConfig.h"
+#include "gpopt/optimizer/COptimizerConfig.h"
+
+using namespace gpopt;
+
+// append given histograms to current object
+void
+CHistogramUtils::AddHistograms
```
Comment (vraghavan78) [inline libnaucrates/src/statistics/CHistogramUtils.cpp]: Just moving.

File: server/src/unittest/dxl/statistics/CCardinalityTestUtils.cpp (hunk 1 above)
Comment (vraghavan78) [inline server/src/unittest/dxl/statistics/CCardinalityTestUtils.cpp]: yes

**Comment by vraghavan78** [comment]
Merged. Thanks!

**Pull request closed without merging**